From 0fcfba5c62581522ad8a5a13340708852df1375a Mon Sep 17 00:00:00 2001
From: Qi Wang <interwq@gwu.edu>
Date: Fri, 19 Mar 2021 22:50:22 -0700
Subject: [PATCH 17/22] Fix locking on arena_i_destroy_ctl().

The ctl_mtx should be held to protect against concurrent arenas.create.

Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
Change-Id: Ifad2388f58f48441eb2fd7049d477cac41e503ca
---
 src/ctl.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/src/ctl.c b/src/ctl.c
index 82af6d96..998c14db 100644
--- a/src/ctl.c
+++ b/src/ctl.c
@@ -2191,6 +2191,8 @@ arena_i_destroy_ctl(tsd_t *tsd, const size_t *mib, size_t miblen, void *oldp,
 	arena_t *arena;
 	ctl_arena_t *ctl_darena, *ctl_arena;
 
+	malloc_mutex_lock(tsd_tsdn(tsd), &ctl_mtx);
+
 	ret = arena_i_reset_destroy_helper(tsd, mib, miblen, oldp, oldlenp,
 	    newp, newlen, &arena_ind, &arena);
 	if (ret != 0) {
@@ -2221,6 +2223,8 @@ arena_i_destroy_ctl(tsd_t *tsd, const size_t *mib, size_t miblen, void *oldp,
 
 	assert(ret == 0);
 label_return:
+	malloc_mutex_unlock(tsd_tsdn(tsd), &ctl_mtx);
+
 	return ret;
 }
 
-- 
2.32.0

